File: javascript/micm/state.js
import { isScalarNumber } from './utils.js';
import { getBackend } from '../backend.js';
import { toWasmSolverType } from './solver.js';
import { GAS_CONSTANT } from './utils.js';

/**
 * Chemical state for one or more grid cells.
 *
 * Holds species concentrations, environmental conditions, and user-defined
 * rate parameters. Create via {@link MICM#createState} rather than directly.
 */
export class State {
  /** @private */
  constructor(nativeMICM, numberOfGridCells, solverType) {
    if (!nativeMICM) {
      throw new TypeError('nativeMICM is required');
    }
    if (numberOfGridCells < 1) {
      throw new RangeError('number_of_grid_cells must be greater than 0');
    }

    const backend = getBackend();
    this._nativeState = backend.create_state(nativeMICM, numberOfGridCells);

    this._numberOfGridCells = numberOfGridCells;
    this._solverType = toWasmSolverType(solverType);
  }

  /**
   * Set species concentrations.
   *
   * @param {Object.<string, number|number[]>} concentrations - Map of species name to
   *   concentration(s) in mol m⁻³. For a single grid cell, values may be scalars.
   *   For multiple grid cells, provide arrays of length `numberOfGridCells`.
   */
  setConcentrations(concentrations) {
    const formatted = {};
    for (const [name, value] of Object.entries(concentrations)) {
      formatted[name] = isScalarNumber(value) ? [value] : value;
    }
    this._nativeState.set_concentrations(formatted, this._solverType);
  }

  /**
   * Get species concentrations for all grid cells.
   *
   * @returns {Object.<string, number[]>} Map of species name to array of concentrations
   *   (one element per grid cell) in mol m⁻³.
   */
  getConcentrations() {
    return this._nativeState.get_concentrations(this._solverType);
  }

  /**
   * Set user-defined rate parameters (e.g. photolysis rates or emission fluxes).
   *
   * @param {Object.<string, number|number[]>} params - Map of parameter name to value(s).
   *   For a single grid cell, values may be scalars. For multiple grid cells, provide
   *   arrays of length `numberOfGridCells`.
   */
  setUserDefinedRateParameters(params) {
    const formatted = {};
    for (const [name, value] of Object.entries(params)) {
      formatted[name] = isScalarNumber(value) ? [value] : value;
    }
    this._nativeState.set_user_defined_constants(formatted, this._solverType);
  }

  /**
   * Get user-defined rate parameters for all grid cells.
   *
   * @returns {Object.<string, number[]>} Map of parameter name to array of values
   *   (one element per grid cell).
   */
  getUserDefinedRateParameters() {
    return this._nativeState.get_user_defined_constants(this._solverType);
  }

  /**
   * Set environmental conditions for each grid cell.
   *
   * All parameters accept a scalar (single grid cell) or an array of length
   * `numberOfGridCells`. If `airDensities` is omitted, air density is computed
   * from the Ideal Gas Law using the provided temperature and pressure.
   *
   * @param {Object} [options]
   * @param {number|number[]|null} [options.temperatures=null] - Temperature(s) in Kelvin
   * @param {number|number[]|null} [options.pressures=null] - Pressure(s) in Pascals
   * @param {number|number[]|null} [options.airDensities=null] - Air number density in mol m⁻³
   */
  setConditions({ temperatures = null, pressures = null, airDensities = null } = {}) {
    const backend = getBackend();
    const vec = new backend.VectorConditions();

    const expand = (param) => {
      if (param === null || param === undefined) {
        return Array(this._numberOfGridCells).fill(null);
      } else if (!Array.isArray(param)) {
        if (this._numberOfGridCells > 1) {
          throw new Error('Scalar input requires a single grid cell');
        }
        return [param];
      } else if (param.length !== this._numberOfGridCells) {
        throw new Error(`Array input must have length ${this._numberOfGridCells}`);
      }
      return param;
    };

    const temps = expand(temperatures);
    const pres = expand(pressures);
    const dens = expand(airDensities);

    for (let i = 0; i < this._numberOfGridCells; i++) {
      const T = temps[i] !== null ? temps[i] : NaN;
      const P = pres[i] !== null ? pres[i] : NaN;
      let rho =
        dens[i] !== null
          ? dens[i]
          : !Number.isNaN(T) && !Number.isNaN(P)
            ? P / (GAS_CONSTANT * T)
            : NaN;

      const cond = new backend.Condition(T, P, rho);
      vec.push_back(cond);
    }

    this._nativeState.set_conditions(vec);
  }

  /**
   * Get environmental conditions for all grid cells.
   *
   * @returns {Array<{temperature: number, pressure: number, air_density: number}>}
   *   One object per grid cell.
   */
  getConditions() {
    return this._nativeState.get_conditions();
  }

  /**
   * Get the number of grid cells in this state.
   * @returns {number}
   */
  getNumberOfGridCells() {
    return this._numberOfGridCells;
  }

  /**
   * Free the underlying WASM object. Call when done with this instance.
   */
  delete() {
    this._nativeState.delete();
  }
}

File: julia/src/micm/state.jl
# Copyright (C) 2023-2026 University Corporation for Atmospheric Research
# SPDX-License-Identifier: Apache-2.0

"""
    State

Chemical state containing species concentrations, environmental conditions,
and user-defined rate parameters for one or more grid cells.

Wraps the C++ `musica::State` object and provides high-level accessors
with proper vector-ordering logic.
"""
mutable struct State
    _ptr::StatePtr
    _micm_ref::Any  # prevent GC of parent MICM before this State
    _species_ordering::Dict{String, Int}
    _rate_parameter_ordering::Dict{String, Int}
    _number_of_grid_cells::Int
    _vector_size::Int

    function State(solver_ptr, number_of_grid_cells::Int, vector_size::Int, micm_ref)
        number_of_grid_cells >= 1 || error("number_of_grid_cells must be >= 1")

        state_ptr = cpp_create_state(solver_ptr, Int64(number_of_grid_cells))

        # Build species ordering dict
        names = cpp_state_species_ordering_names(state_ptr)
        indices = cpp_state_species_ordering_indices(state_ptr)
        species_ordering = Dict{String, Int}()
        for i in 1:length(names)
            species_ordering[names[i]] = Int(indices[i])
        end

        # Build rate parameter ordering dict
        rnames = cpp_state_rate_param_ordering_names(state_ptr)
        rindices = cpp_state_rate_param_ordering_indices(state_ptr)
        rate_ordering = Dict{String, Int}()
        for i in 1:length(rnames)
            rate_ordering[rnames[i]] = Int(rindices[i])
        end

        obj = new(state_ptr, micm_ref, species_ordering, rate_ordering, number_of_grid_cells, vector_size)
        finalizer(obj) do s
            cpp_delete_state(s._ptr)
        end
        return obj
    end
end

"""
    set_concentrations!(state::State, concentrations::Dict{String})

Set species concentrations. Values can be scalars (for single grid cell)
or vectors matching the number of grid cells.
"""
function set_concentrations!(state::State, concentrations::Dict{String, <:Any})
    n_species = length(state._species_ordering)
    for (name, value) in concentrations
        haskey(state._species_ordering, name) || error("Species $name not found in the mechanism.")
        i_species = state._species_ordering[name]
        values = value isa Real ? [Float64(value)] : Float64.(value)
        length(values) == state._number_of_grid_cells || error(
            "Concentration list for $name must have length $(state._number_of_grid_cells).")
        for i_cell in 1:state._number_of_grid_cells
            idx = flat_index(i_cell, i_species, n_species, state._vector_size)
            cpp_state_set_concentration!(state._ptr, Int64(idx), values[i_cell])
        end
    end
end

"""
    get_concentrations(state::State) -> Dict{String, Vector{Float64}}

Get species concentrations for all grid cells.
"""
function get_concentrations(state::State)
    concentrations = Dict{String, Vector{Float64}}()
    n_species = length(state._species_ordering)
    for (species, i_species) in state._species_ordering
        vals = Vector{Float64}(undef, state._number_of_grid_cells)
        for i_cell in 1:state._number_of_grid_cells
            idx = flat_index(i_cell, i_species, n_species, state._vector_size)
            vals[i_cell] = cpp_state_get_concentration(state._ptr, Int64(idx))
        end
        concentrations[species] = vals
    end
    return concentrations
end

"""
    set_user_defined_rate_parameters!(state::State, params::Dict{String})

Set user-defined rate parameters. Values can be scalars (for single grid cell)
or vectors matching the number of grid cells.
"""
function set_user_defined_rate_parameters!(state::State, params::Dict{String, <:Any})
    n_params = length(state._rate_parameter_ordering)
    for (name, value) in params
        haskey(state._rate_parameter_ordering, name) || error(
            "User-defined rate parameter $name not found in the mechanism.")
        i_param = state._rate_parameter_ordering[name]
        values = value isa Real ? [Float64(value)] : Float64.(value)
        length(values) == state._number_of_grid_cells || error(
            "User-defined rate parameter list for $name must have length $(state._number_of_grid_cells).")
        for i_cell in 1:state._number_of_grid_cells
            idx = flat_index(i_cell, i_param, n_params, state._vector_size)
            cpp_state_set_rate_param!(state._ptr, Int64(idx), values[i_cell])
        end
    end
end

"""
    get_user_defined_rate_parameters(state::State) -> Dict{String, Vector{Float64}}

Get user-defined rate parameters for all grid cells.
"""
function get_user_defined_rate_parameters(state::State)
    result = Dict{String, Vector{Float64}}()
    n_params = length(state._rate_parameter_ordering)
    for (param, i_param) in state._rate_parameter_ordering
        vals = Vector{Float64}(undef, state._number_of_grid_cells)
        for i_cell in 1:state._number_of_grid_cells
            idx = flat_index(i_cell, i_param, n_params, state._vector_size)
            vals[i_cell] = cpp_state_get_rate_param(state._ptr, Int64(idx))
        end
        result[param] = vals
    end
    return result
end

"""
    set_conditions!(state::State; temperatures, pressures, air_densities)

Set environmental conditions for each grid cell. For a single grid cell,
scalar values are accepted. If `air_densities` is not provided, it is
calculated from the Ideal Gas Law.
"""
function set_conditions!(state::State;
    temperatures::Union{Real, AbstractVector{<:Real}, Nothing}=nothing,
    pressures::Union{Real, AbstractVector{<:Real}, Nothing}=nothing,
    air_densities::Union{Real, AbstractVector{<:Real}, Nothing}=nothing,
)
    n = state._number_of_grid_cells

    function expand(param, name)
        if param === nothing
            return nothing
        elseif param isa Real
            n == 1 || error("$name must be a vector of length $n for multi-cell states.")
            return [Float64(param)]
        else
            length(param) == n || error("$name must have length $n.")
            return Float64.(param)
        end
    end

    temps = expand(temperatures, "temperatures")
    press = expand(pressures, "pressures")
    dens = expand(air_densities, "air_densities")

    for i_cell in 1:n
        idx = Int64(i_cell - 1)  # 0-based for C++
        cur_temp = temps !== nothing ? temps[i_cell] : cpp_state_get_condition_temperature(state._ptr, idx)
        cur_pres = press !== nothing ? press[i_cell] : cpp_state_get_condition_pressure(state._ptr, idx)
        if dens !== nothing
            cur_dens = dens[i_cell]
        elseif cur_temp > 0.0 && cur_pres > 0.0
            cur_dens = cur_pres / (GAS_CONSTANT * cur_temp)
        else
            cur_dens = 0.0
        end
        cpp_state_set_condition!(state._ptr, idx, cur_temp, cur_pres, cur_dens)
    end
end

"""
    get_conditions(state::State) -> Dict{String, Vector{Float64}}

Get environmental conditions for all grid cells.
Returns a dictionary with keys "temperature", "pressure", "air_density".
"""
function get_conditions(state::State)
    temps = Vector{Float64}(undef, state._number_of_grid_cells)
    press = Vector{Float64}(undef, state._number_of_grid_cells)
    dens = Vector{Float64}(undef, state._number_of_grid_cells)
    for i_cell in 1:state._number_of_grid_cells
        idx = Int64(i_cell - 1)
        temps[i_cell] = cpp_state_get_condition_temperature(state._ptr, idx)
        press[i_cell] = cpp_state_get_condition_pressure(state._ptr, idx)
        dens[i_cell] = cpp_state_get_condition_air_density(state._ptr, idx)
    end
    return Dict{String, Vector{Float64}}(
        "temperature" => temps,
        "pressure" => press,
        "air_density" => dens,
    )
end

"""
    get_species_ordering(state::State) -> Dict{String, Int}

Get the mapping of species names to their indices.
"""
get_species_ordering(state::State) = state._species_ordering

"""
    get_user_defined_rate_parameters_ordering(state::State) -> Dict{String, Int}

Get the mapping of user-defined rate parameter names to their indices.
"""
get_user_defined_rate_parameters_ordering(state::State) = state._rate_parameter_ordering
